---
title: "GeneGraphs"
output:
  html_document: default
---
## Purpouse 

This script processes raw ribosome profiling data in order to visualize mapping of ribosome footprints to the transcriptome. The purpouse is to identify  alternative upstream translation initiation and quantify it in respect to canonical initiation.


```{r the set up}
#Loading libraries

library(rhdf5)
library(tidyr)
library(dplyr)
library(magrittr)
library(purrr)
library(ggplot2)
library(tibble)

#Set up

nnt_gene<- 50
setwd("/Users/Ania/Desktop/SzkoÅ‚a/4th year/Dissertation/gene_graphs/")

dataset <- "G-Sc_2014"
hd_file <- "G-Sc_2014/output/WTnone/WTnone.h5"

# prepare files, opens hdf5 file connection
hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
# formal class H5IDComponent

gene_names <- rhdf5::h5ls(hdf5file, recursive = 1)$name
#  chr [1:5812] "Q0045" "Q0050"

# next step is to test with these 5 genes only instead of all genes
test_orfs<- c("YCR012W","YEL009C","YOR303W","YOL130W","YGR094W")

orf_gff_file <- "G-Sc_2014/input/yeast_CDS_w_250utrs.gff3"
# chr [1:5] "YCR012W" "YEL009C"

#using function readGFFAsDf (defined in functions)
gff_df <- readGFFAsDf(orf_gff_file)

```

##Functions

``` {r}

# Creates a matrix with number of columns that start at 5' UTR and finish at 50'th nt of the CDS
GetGeneDatamatrix5UTR <- function(gene, dataset, hdf5file, utr5_data, nnt_gene) {
  data_mat_all <- GetGeneDatamatrix(gene, dataset, hdf5file)
  #I'm changing n_left5 to just 1st nt of 5'UTR
  n_left5 <- Get5UTRstart(gene, utr5_data) # column to start from (5'end)
  n_right3 <- UTR5_length(gene, utr5_data) + nnt_gene # column to end with (3'end) 
  data_mat_5start <- data_mat_all[, n_left5 : n_right3]
  posn5start <- UTR5_5start(gene, utr5_data) # get utr5 start position (e.g. -250)
  posn5end <- CDS3_end(gene, utr5_data) # get utr5 end position (e.g. -1) + nnt_gene
  data_mat_5start <- tibble::as_tibble(data_mat_5start, .name_repair="minimal")
  names(data_mat_5start) <- as.character(seq(from=posn5start, to=posn5end)) # add position as character 'name's to columns
  data_mat_5start <-tibble::add_column(data_mat_5start, read_length=10:50, .before=1) # minreadlength:maxreadlength
  return(data_mat_5start)
  #return(data_mat_5start, posn5start, posn5end)
}

#Used within GetGeneDataMatrix5UTR to create a matrix from hdf5file
GetGeneDatamatrix <- function(gene, dataset, hdf5file) {
  hdf5file %>%
    rhdf5::H5Dopen(
      name = paste0("/", gene, "/", dataset, "/reads/data")
    ) %>%
    rhdf5::H5Dread() %>%
    return()
}

#Modifies GFF table: 
  #changes 5UTR start from 1 to -250
  #changes 5UTR end from 250 to -1
modifyTo_5UTR_table <- function(x){
  x %>%
    filter(type=="UTR5", strand == "+") %>%
    # # A tibble: 5,812 x 10
    mutate(utr5start = (end * -1), utr5end = (start -2))  # FLIC: should this be -2 or -1?
}

# Used within GetGeneDatamatrix5UTR: takes value for start position of 5'UTR for matrix creation (no negative values e.g.: 1)
Get5UTRstart <- function(gene, x) {
  x %>% 
    dplyr::filter(Name == gene) %>% 
    dplyr::pull(start)
}

# Used within GetGeneDatamatrix5UTR: takes value for length of 5'UTR  
UTR5_length <- function(gene, x) {
  x %>% 
    dplyr::filter(Name == gene) %>% 
    dplyr::pull(width)
} 

# Used within GetGeneDatamatrix5UTR: takes value for start position of 5'UTR for naming matrix col's (:-250)
UTR5_5start <- function(gene, x) {
  x %>% 
    dplyr::filter(Name == gene) %>% 
    dplyr::pull(utr5start)
} 


# Used within GetGeneDatamatrix5UTR: Takes value for 3' end of the table (includes whole 5'UTR +50 nt of CDS)
CDS3_end <- function(gene, x) {
  x %>% 
    dplyr::filter(Name == gene) %>% 
    dplyr::pull(utr5end) + nnt_gene
} 
```


##Modifications of the GFF table

```{r}
#GFF initially
gff_df

```


``` {r}
utr5_data <- modifyTo_5UTR_table(gff_df)
head(utr5_data)
```
Columns start and end are used to establish number of columns in the matrix (col 1:300). 
Columns utr5start and utr5end are used to name columns with biologically accurate names for positions (col -250:50--> -250 is 5'UTR start and 50 is first 50nt of CDS). 

## The GetGeneDatamatrix5UTR
``` {r}
#for 1 gene
utr5Matrix_1gene <- GetGeneDatamatrix5UTR(test_orfs[1], dataset, hdf5file, utr5_data, nnt_gene)


```


``` {r}
# for test_orfs (5 genes)
utr5Matrix_allGenes <- purrr::map(test_orfs, GetGeneDatamatrix5UTR, dataset, hdf5file, utr5_data, nnt_gene)

# apply the names to each item in the list
names(utr5Matrix_allGenes) <- test_orfs

```

## Tidying data

``` {r tidying data}

GetPosnCountOutput <- function(x){
  output_thing <- x %>% 
    gather(-read_length, key="Position", value = "Counts") %>%
    group_by(Position) %>%
    summarise(Counts=sum(Counts)) %>%
    arrange(as.integer(Position))
  return(output_thing)
}

PosnCountAllgenes <- purrr::map(utr5Matrix_allGenes, GetPosnCountOutput)

```

## Plotting 

Plotting with ggplot.

```{r plot bit}
#That's for a single gene for now (previous attempts were done on 1 gene only)
try_plot<-ggplot(
  data=try,) +
  geom_histogram(aes(x=as.numeric(Position), y=Counts), stat="identity") + 
  theme_light() +
  labs(y= "Read count", x = "Position (mRNA)") +
  scale_x_discrete(limit = c(-100,-50,0,50))

try_plot

ggsave("try_plot_example", device = "png")

## Multi-facet 

 mplot<-ggarrange(x1_plot, x2_plot, x3_plot, x4_plot, x5_plot, 
           labels = c("A", "B", "C", "D", "E"),
           ncol = 2, nrow = 3)

```